py/asmrv32.c

```diff
@@ -354,31 +364,25 @@ void asm_rv32_emit_jump_if_reg_eq(asm_rv32_t *state, mp_uint_t rs1, mp_uint_t rs
 }
 
 void asm_rv32_emit_jump_if_reg_nonzero(asm_rv32_t *state, mp_uint_t rs, mp_uint_t label) {
-    ptrdiff_t displacement = (ptrdiff_t)(state->base.label_offsets[label] - state->base.code_offset);
+    ptrdiff_t displacement = 0;
+    bool emit_short_jump = false;
+    calculate_displacement_for_label(state, label, &displacement, &emit_short_jump);
 
-    if (FIT_SIGNED(displacement, 9) && IS_IN_C_REGISTER_WINDOW(rs)) {
+    if (FIT_SIGNED(displacement, 8) && emit_short_jump && IS_IN_C_REGISTER_WINDOW(rs)) {
```

Comment:
I suggest to put the `emit_short_jump` (or `can_emit_short_jump`) first in the if-logic, because it's the most important thing to check first, whether this optimisation can be done or not.  It'll also be a little more efficient when executing.

Reply:
I've implemented your suggestion, but the compiler emitted the very same code when optimisations were enabled as `can_emit_short_jump` is the least computationally-intensive thing to check for if you need to invalidate the whole expression (a non-volatile boolean), so it was checked first anyway.

Reply:
Oh, that's interesting that the compiler swaps the operands of the short-circuiting operator.  I guess it can do that if it's all local variables and arithmetic (no function calls).

Anyway, thanks for updating it.